// @bhtbot/bhtbotservice/build/main/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayAnswer = exports.GatewayRequestEntity = exports.GatewayRequest = exports.DefaultMiddleware = exports.AppConfig = exports.Service = void 0;
const AppConfig_1 = __importDefault(require("./lib/AppConfig"));
exports.AppConfig = AppConfig_1.default;
const service_1 = __importDefault(require("./lib/service"));
exports.Service = service_1.default;
const GatewayRequest_1 = __importDefault(require("./lib/GatewayRequest"));
exports.GatewayRequest = GatewayRequest_1.default;
const GatewayRequestEntity_1 = __importDefault(require("./lib/GatewayRequestEntity"));
exports.GatewayRequestEntity = GatewayRequestEntity_1.default;
const GatewayAnswer_1 = __importDefault(require("./lib/GatewayAnswer"));
exports.GatewayAnswer = GatewayAnswer_1.default;
const DefaultMiddleware = __importStar(require("./lib/DefaultMiddleware"));
exports.DefaultMiddleware = DefaultMiddleware;
