// @bhtbot/bhtbot/build/main/lib/botRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class BotRequest {
    constructor(payload) {
        if (payload === undefined ||
            payload === null ||
            payload.text === undefined ||
            payload.text === null) {
            throw new Error('Payload must contain text');
        }
        this.payload = payload;
        this.computeValues();
        console.log('Bot Request with payload', this.payload);
    }
    computeValues(serviceId) {
        serviceId = serviceId !== null && serviceId !== void 0 ? serviceId : BotRequest.serviceId;
        this.payload.serviceName = serviceId;
    }
    setText(text) {
        this.payload.text = text;
        return this;
    }
    setUserId(userId) {
        if (userId.toString().indexOf('@') > 0) {
            //strip off service name if provided
            userId = userId.toString().split('@')[0];
        }
        this.payload.serviceUserId = userId;
        return this;
    }
    setClientLanguage(language) {
        this.payload.clientLanguage = language;
        return this;
    }
    setUserTime(clientDate) {
        this.payload.clientDate = clientDate;
        return this;
    }
    setUserFirstname(firstname) {
        this.payload.firstName = firstname;
        return this;
    }
    get userId() {
        return this.payload.serviceUserId + '@' + this.payload.serviceName;
    }
}
exports.default = BotRequest;
